# $$Marketing~Analytics:Pesquisa~Operacional$$

# $$Otimização~de~Preços~e~Mix~de~Produtos$$

## $$Conceitos:$$


**É um método analítico avançado que permite a solução de problemas e a tomada de decisões nas organizações**


**Pilares para tomada de decisões assertivas:**

- Método Cientifica
- Modelagem de sistemas 
- Abordagem Interdisciplinar 
- Uso de Computadores
- Pragmatismo



**Modelos da P.O:**

- Programação Matemática (P.Linear, P.Inteira)
- Machine Learning
- Simulação a Eventos Discretos
- Apoio Multicritério à Decisão
- Teoria dos Grafos
- Lógica Fuzzy
- Estatística
- Análise Envoltória de Dados

**Fatores que influenciaram a Expansão da P.O:**

- Novos modelos matemáticos
- Evolução dos computadores


**O QUE É UM SISTEMA:**

- É um conjunto de elementos interdependentes que interagem com objetivos comuns. Normalmente, o resultado gerado pelo sistema é maior do que o somatório dos resultados isolados das partes que compõem o sistema

**MEDIDA DE EFICÁCIA OPERACIONAL(MEO):**

- É uma medida, quantificável, que procura representar o grau com que dado sistema, em condições de operação tão reais quanto possível, atinge sua finalidade. A MEO permite quantificar aspectos pertinentes a requisitos operacionais específicos


- **Três atributos de uma boa MEO:**

- Quantificável
- Mensurável
- “Closely Related”


**MODELO:**

- É uma representação simplificada da realidade, destinada a facilitar o entendimento e, consequentemente, a manipulação de um fenômeno complexo


**ETAPAS PARA SOLUÇÃO DE PROBLEMAS:**

- 1° - Entender o sistema (objetivo)
- 2° - Definir as MEO
- 3° - Construir o Modelo
- 4° - Obter uma massa de dados
- 5° - Melhorar o modelo
- 6° - Validar o modelo
- 7° - Reportar, explicar, comunicar e auxiliar na implementação do modelo (STORYTELLING)


**P.O. E A CIÊNCIA DE DADOS:**

- Pesquisa Operacional e Ciência de Dados estão intimamente relacionadas. Os modelos da P.O. precisam dos dados para que esses dados sejam processados e possam gerar um resultado útil para o tomador de decisão. Por outro lado, de nada adianta dispor de um conjunto de dados, em quantidade e com qualidade, se não houver um modelo matemático que possa dar forma a esses dados, de maneira a apoiar adequadamente o processo de tomada de decisão

**BRAINSTORMING:**

- “tempestade de ideias”, aborda a utilização do processo criativo para a solução de problemas.


**MAPA MENTAL:**

- Um mapa mental é uma ferramenta simples utilizada para organizar o pensamento. Um mapa mental é uma maneira fácil de extrair informações do cérebro humano


- Com um mapa mental, uma longa lista de informações pode se transformar num diagrama colorido, fácil de lembrar, intuitivo e bem organizado que opera em harmonia com o funcionamento natural do cérebro



## $$Conceito~de~Problema~de~Programação~Linear (PPL)$$

- Maximização ou Minimização de uma Função Objetivo (F.O.) linear com relação às Variáveis de Decisão (V.D.) do modelo.


- Respeitando-se as restrições do problema expressas por um sistema de equações e/ou inequações lineares associadas com as Variáveis de Decisão do modelo


- VARIÁVEIS DE DECISÃO: são aqueles valores que representam o cerne do problema. São as quantidades pelas quais devemos decidir


- As variáveis de decisão representam as opções que um gestor tem para atingir um objetivo


- SUPOSIÇÕES DA PL: Certeza, Divisibilidade, Aditividade e Proporcionalidade


**SOLUÇÃO DE UM PPL:**

Qualquer especificação de valores para as variáveis de decisão é chamada solução, independentemente de ela ser desejável ou não


Quando o problema envolve apenas duas variáveis de decisão, a solução ótima de um problema de programação linear pode ser encontrada graficamente.


- **solução viável** = É aquela para a qual todas as restrições são satisfeitas


- **Inviável** = É aquela para a qual pelo menos uma das restrições não é satisfeita


- **solução ótima** = É aquela em que todas as restrições são atendidas, otimizando-se o valor da função objetivo.


**PPL NA FORMA PADRÃO:**

- A Função-Objetivo é de Maximização


- As restrições têm sinal de menor ou igual


- As constantes de todas as restrições são não-negativas


- As variáveis de decisão só podem assumir valores não-negativos


**Variáveis de Folga ou Excesso:**

- Dado um Problema de Programação Linear, precisamos inserir as variáveis de folga ou excesso para que as desigualdades se tornem igualdades

__------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------__


__------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------__

## $$Desafio~ de~ PO:$$

**A empresa quer montar e testar dois modelos de carros, XTZ1 e XTZ2. Para o próximo mês, a empresa quer decidir quantas unidades de cada modelo vai montar e depois testar**

**Nenhum Carro está em estoque, como esses modelos serão limitados somente para este mês, a empresa não quer manter nenhum estoque para o mês seguinte**

**Ela acredita que o máximo que pode vender neste mês são 300 unidades do modelo XTZ1 e 150 unidades do modelo XTZ2**

**Cada modelo XTZ1 é vendido por RS300.000,00 e cada modelo Lua2 por RS370.000,00 O custo dos componentes de um XTZ1 é de RS150.000,00 e para um XTZ2 é RS185.000,00**

**A mão de obra é necessária para a montagem e teste. Existem no máximo 5.000 horas de montagem e 3.000 horas de teste disponíveis. Cada hora de trabalho para montagem custa RS110 e cada hora de trabalho para teste custa RS150. Cada carro XTZ1 requer doze horas para montagem e uma hora para teste. Cada XTZ2 requer quatorze horas para montagem e duas horas para teste**

**A Empresa deseja saber quantas unidades de cada modelo deve produzir (montar e testar) para maximizar seu lucro líquido, mas não pode usar mais horas de trabalho do que as disponíveis e não deseja produzir mais do que pode vender**




__------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------__


__------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------__

## $$Objetivo:$$  

## $$Realizar~a ~otimização~de~preços~e~ mix ~dos~produtos~da~empresa$$

In [1]:
# Pacote
from pulp import *

## Criando o Modelo Matemático Para a Otimização

### Parâmetros

- Ai = Número máximo de carros modelo tipo i para vender este mês, onde i pertence ao conjunto {XTZ1, XTZ2}


- Bi = Preço de venda de carros modelo tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}


- Ci = Preço de custo das peças componentes para os carros modelo tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}


- Di = Custo de mão de obra de montagem por hora dos carros modelo tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}


- Ei = Custo de mão de obra de teste por hora dos carros modelo tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}


- F = Número máximo de horas de trabalho de montagem


- G = Número máximo de horas de trabalho de teste


- Hf,i = Horas de montagem necessárias para construir cada modelo de carro tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}


- Hg,i = Horas de teste necessárias para testar cada modelo de carro tipo i, onde i pertence ao conjunto {XTZ1, XTZ2}

**Para criar o modelo matemático, primeiro devemos entender o problema de negócio para assim construir o modelo de otimização**

### Variável de Decisão

- Xi = Número de carros modelo tipo i a produzir este mês, onde i pertence ao conjunto {XTZ1, XTZ2}

**A Variavel de decisão é o resultado do modelo matemático, variavel resposta**

### Função Objetivo

![title](imagens/formula.png)

**Lucro total de cada empresa depende de varios fatores que contribuem para os lucros e gastos, estamos simplificando a formula do lucro total apenas para fins didáticos**

### Restrições

- O número de carros do modelo tipo i a serem produzidos não pode ser negativo, ou seja, Xi >= 0, onde i pertence ao conjunto {XTZ1, XTZ2}.


- O número total de horas de montagem não pode ser maior que o número máximo de horas de mão de obra de montagem disponíveis.


- O número total de horas de teste não pode ser maior do que o máximo de horas de mão de obra de teste disponíveis.


- O número de carros do modelo tipo i a serem produzidos não pode ser maior do que o número máximo de carros do modelo tipo i a serem vendidos neste mês, onde i pertence ao conjunto {XTZ1, XTZ2}.

**As restrições também variam dependendo do projeto e do objetivo que a empresa quer atingir, criamos algumas restrições que fazem sentido para o problema proposto**

## Implementando o Modelo Matemático

### Organizando os Parâmetros

In [2]:
# Número máximo de Carros para vender este mês
A_XTZ1 = 300
A_XTZ2 = 150

In [3]:
# Preço de venda dos Carros 
B_XTZ1 = 300000
B_XTZ2 = 370000

In [4]:
# Preço de custo das peças componentes para os carros
C_XTZ1 = 150000
C_XTZ2 = 185000

In [5]:
# Custo de mão de obra de montagem por hora dos carros
D_XTZ1 = 110
D_XTZ2 = 110

In [6]:
# Custo de mão de obra de teste por hora dos carros
E_XTZ1 = 150
E_XTZ2 = 150

In [7]:
# Número máximo de horas de trabalho de montagem
F = 5000

In [8]:
# Número máximo de horas de trabalho de teste
G = 3000

In [9]:
# Horas de montagem necessárias para construir cada modelo de carro
Hfi_XTZ1 = 12
Hfi_XTZ2 = 14

In [10]:
# Horas de teste necessárias para testar cada modelo de carro
Hgi_XTZ1 = 1
Hgi_XTZ2 = 2

**A organização dos parâmetros é o processo onde colocamos todas as informações numéricas da empresa em variáveis para chegar no objetivo desejado**

### Criação da Variável Para o Problema de Otimização

In [11]:
# Criando a variável para o problema
problema = LpProblem("MixCarros", LpMaximize)

In [12]:
# Visualiza
problema

MixCarros:
MAXIMIZE
None
VARIABLES

### Definindo a Variável de Decisão para Cada Modelo dos Carros

In [13]:
# Define as variáveis
x_XTZ1 = LpVariable("Unidades XTZ1", 0, None, LpInteger)
x_XTZ2 = LpVariable("Unidades XTZ2", 0, None, LpInteger)

In [14]:
# Print
print(x_XTZ1)
print(x_XTZ2)

Unidades_XTZ1
Unidades_XTZ2


**Definimos que cada modelo de carro pode ter de 0 carros fabricados no mês até o número definido nos parâmetros acima**

### Implementando a Função Objetivo

![title](imagens/formula.png)

## Replicando a formula acima:

In [15]:
# Faturamento
faturamento = (x_XTZ1 * B_XTZ1) + (x_XTZ2 * B_XTZ2)
faturamento

300000*Unidades_XTZ1 + 370000*Unidades_XTZ2 + 0

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados



- **B_XTZ1 e B_XTZ2:** Preço de venda dos Carros 

In [16]:
# Custo de Montagem
custo_montagem = (x_XTZ1 * Hfi_XTZ1 * D_XTZ1) + (x_XTZ2 * Hfi_XTZ2 * D_XTZ2)
custo_montagem

1320*Unidades_XTZ1 + 1540*Unidades_XTZ2 + 0

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados



- **Hfi_XTZ1 e Hfi_XTZ2:** Horas de montagem necessárias para construir cada modelo de carro



- **D_XTZ1 e D_XTZ2:** Custo de mão de obra de montagem por hora dos carros

In [17]:
# Custo de Teste
custo_teste = (x_XTZ1 * Hgi_XTZ1 * E_XTZ1) + (x_XTZ2 * Hgi_XTZ2 * E_XTZ2)
custo_teste

150*Unidades_XTZ1 + 300*Unidades_XTZ2 + 0

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados



- **Hgi_XTZ1 e Hgi_XTZ2:** Horas de teste necessárias para testar cada modelo de carro



- **E_XTZ1 e E_XTZ2:** Custo de mão de obra de teste por hora dos carros

In [18]:
# Custo de Componentes
custo_componentes = (x_XTZ1 * C_XTZ1) + (x_XTZ2 * C_XTZ2) 
custo_componentes

150000*Unidades_XTZ1 + 185000*Unidades_XTZ2 + 0

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados



- **C_XTZ1 e C_XTZ2:** Preço de custo das peças componentes para os carros 

In [19]:
# Lucro = Faturamento - Custo de Montagem - Custo de Teste - Custo de Componentes
problema += faturamento - custo_montagem - custo_teste - custo_componentes
problema

MixCarros:
MAXIMIZE
148530*Unidades_XTZ1 + 183160*Unidades_XTZ2 + 0
VARIABLES
0 <= Unidades_XTZ1 Integer
0 <= Unidades_XTZ2 Integer

**Formula para encontra o lucro**

### Restrições

In [20]:
# Número máximo de horas de montagem
problema += (x_XTZ1 * Hfi_XTZ1) + (x_XTZ2 * Hfi_XTZ2) <= F,"Número Máximo de Horas de Montagem"

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados



- **Hfi_XTZ1 e Hfi_XTZ2:** Horas de montagem necessárias para construir cada modelo de carro

In [21]:
# Número máximo de horas de teste
problema += (x_XTZ1 * Hgi_XTZ1) + (x_XTZ2 * Hgi_XTZ2) <= G,"Número Máximo de Horas de Teste"

**Variáveis usadas:**

- **x_XTZ1 e x_XTZ2:** Quantidade de carros a serem fabricados




- **Hgi_XTZ1 e Hgi_XTZ2:** Horas de teste necessárias para testar cada modelo de carro

In [22]:
# Produção menor ou igual a demanda pelo modelo 1
problema += x_XTZ1 <= A_XTZ1,"Produção menor ou igual a demanda pelo modelo 1"

**Variáveis usadas:**

- **x_XTZ1:** Quantidade de carros a serem fabricados, modelo 1




- **A_XTZ1:** Número máximo de Carros para vender este mês, modelo 1

In [23]:
# Produção menor ou igual a demanda pelo modelo 2
problema += x_XTZ2 <= A_XTZ2,"Produção menor ou igual a demanda pelo modelo 2"

**Variáveis usadas:**

- **x_XTZ2:** Quantidade de carros a serem fabricados, modelo 2




- **A_XTZ2:** Número máximo de Carros para vender este mês, modelo 2

In [24]:
# Problema final
problema

MixCarros:
MAXIMIZE
148530*Unidades_XTZ1 + 183160*Unidades_XTZ2 + 0
SUBJECT TO
Número_Máximo_de_Horas_de_Montagem: 12 Unidades_XTZ1 + 14 Unidades_XTZ2
 <= 5000

Número_Máximo_de_Horas_de_Teste: Unidades_XTZ1 + 2 Unidades_XTZ2 <= 3000

Produção_menor_ou_igual_a_demanda_pelo_modelo_1: Unidades_XTZ1 <= 300

Produção_menor_ou_igual_a_demanda_pelo_modelo_2: Unidades_XTZ2 <= 150

VARIABLES
0 <= Unidades_XTZ1 Integer
0 <= Unidades_XTZ2 Integer

**Todo o modelo de otimização está dentro da variável 'problema'**

### Resolvendo o Problema de Otimização

In [25]:
# Otimização
problema.solve()

1

In [26]:
# Lucro Maximizado
print("Lucro Maximizado:", value(problema.objective))

Lucro Maximizado: 63349000.0


In [27]:
# Número de Unidades do Modelo XTZ1 a Produzir
print("Número de Unidades do Modelo XTZ1 a Produzir:", problema.variables()[0].varValue)

Número de Unidades do Modelo XTZ1 a Produzir: 244.0


In [28]:
# Número de Unidades do Modelo XTZ2 a Produzir
print("Número de Unidades do Modelo XTZ2 a Produzir:", problema.variables()[1].varValue)

Número de Unidades do Modelo XTZ2 a Produzir: 148.0


**A empresa deve produzir 244 unidades de carros do tipo XTZ1 e 148 unidades do tipo XTZ2 para atingir o lucro máximo de R$63.349.000.00**

##  Conclusão

**Nem todo processo de um cientista de dados vai precisar de uma tabela ou do uso de machine learning para conclusão do projeto, o objetivo é sempre resolver problemas de negócio e esse é mais um caso onde foi resolvido o problema, porém usando outras ferramentas**